plugins/inventory/xen_orchestra.py

```diff
@@ -0,0 +1,303 @@
+            self.inventory.set_variable(
+                host, 'os_version', vm["os_version"])
+
+    def _add_hosts(self, hosts, pools):
+        for host in hosts.values():
+            address = host['address']
+            group_name = self.to_safe("xo_host_{0}".format(host['name_label']))
+            pool_name = self._pool_group_name_for_uuid(pools, host['$poolId'])
+
+            self.inventory.add_group(group_name)
+            self.inventory.add_host(address)
+            self.inventory.add_child(HOST_GROUP, address)
+            self.inventory.add_child(pool_name, address)
+
+            self.inventory.set_variable(address, 'enabled', host['enabled'])
+            self.inventory.set_variable(address, 'hostname', host['hostname'])
+            self.inventory.set_variable(address, 'memory', host['memory'])
+            self.inventory.set_variable(address, 'cpus', host['cpus'])
+            self.inventory.set_variable(address, 'type', 'host')
+            self.inventory.set_variable(address, 'tags', host['tags'])
+            self.inventory.set_variable(address, 'version', host['version'])
+            self.inventory.set_variable(
+                address, 'power_state', host['power_state'].lower())
+            self.inventory.set_variable(
+                address, 'product_brand', host['productBrand'])
+
+        for pool in pools.values():
+            group_name = self.to_safe("xo_pool_{0}".format(pool['name_label']))
+
+            self.inventory.add_group(group_name)
+
+    def _add_pools(self, pools):
+        for pool in pools.values():
+            group_name = self.to_safe("xo_pool_{0}".format(pool['name_label']))
+
+            self.inventory.add_group(group_name)
+
+    # TODO: Refactor
+    def _pool_group_name_for_uuid(self, pools, pool_uuid):
+        for pool in pools:
+            if pool == pool_uuid:
+                return self.to_safe("xo_pool_{0}".format(pools[pool_uuid]['name_label']))
+
+    # TODO: Refactor
+    def _host_group_name_for_uuid(self, hosts, host_uuid):
+        for host in hosts:
+            if host == host_uuid:
+                return self.to_safe("xo_host_{0}".format(hosts[host_uuid]['name_label']))
+
+    def _populate(self, objects):
+        # Prepare general groups
+        self.inventory.add_group(HOST_GROUP)
+        self.inventory.add_group(POOL_GROUP)
+        for group in POWER_STATES:
+            self.inventory.add_group(group.lower())
+
+        self._add_pools(objects['pools'])
+        self._add_hosts(objects['hosts'], objects['pools'])
+        self._add_vms(objects['vms'], objects['hosts'], objects['pools'])
+
+    def to_safe(self, word):
+        '''Converts 'bad' characters in a string to underscores so they can be used as Ansible groups
+        #> InventoryModule.to_safe("foo-bar baz")
+        'foo_barbaz'
+        '''
+        regex = r"[^A-Za-z0-9\_]"
+        return re.sub(regex, "_", word.replace(" ", "")).lower()
```

Comment:
Ansible already has a built-in (and configurable) mechanism to sanitize group names. Have you considered not doing this in this plugin, but simply let Ansible deal with it?

Reply:
I didn't know. I'm touching the tests up and I'll test this as well. Thanks for the review/heads up

Reply:
I'm not relying on built-in sanitizer, although I add to replace `_` and ` ` (spaces) before running them through